mpi-proxy-split/record-replay.cpp

```diff
@@ -670,16 +814,85 @@ setCartesianCommunicator(void *getCartesianCommunicatorFptr)
+  int comm_cart_ranks_order[MAX_PROCESSES];
+
+  CartesianInfo ci;
   CartesianProperties cp;
   CartesianInfo checkpoint_mapping[MAX_PROCESSES];
   CartesianInfo restart_mapping[MAX_PROCESSES];
 
-  load_checkpoint_cartesian_mapping(checkpoint_mapping, cp.comm_old_size,
-                                    cp.ndims);
-*/
+  // In current implementation, <comm_old> is MPI_COMM_WORLD
+  comm_old_prime = MPI_COMM_WORLD;
+
+  // Get cartesian info of this process
+  retval = MPI_Comm_rank(comm_old_prime, &ci.comm_old_rank);
+  retval = MPI_Comm_rank(*comm_cart_prime, &ci.comm_cart_rank);
+
+  // Get cartesian properties of this process
+  dmtcp::ostringstream o;
+  o << "./ckpt_rank_" << ci.comm_old_rank << "/cartesian.info";
+  retval = load_cartesian_properties(o.str().c_str(), &cp);
+  JASSERT(retval == 0)
+  (o.str().c_str()).Text("Failed to load cartesian properties.");
+
+  // Get coordinates of this process
+  retval = MPI_Cart_coords(*comm_cart_prime, ci.comm_cart_rank, cp.ndims,
+                           ci.coordinates);
+
+  // Load checkpoint cartesian mapping
+  load_checkpoint_cartesian_mapping(&cp, checkpoint_mapping);
+
+  // Load restart cartesian mapping
+  load_restart_cartesian_mapping(&cp, &ci, restart_mapping);
+
+  retval = MPI_Barrier(MPI_COMM_WORLD);
+  JASSERT(retval == MPI_SUCCESS).Text("MPI_Barrier(1) failed.");
+
+  // while (dummy) {};
+
+  // The root process will populate <comm_old_ranks_order> and
+  // <comm_cart_ranks_order> arrays
+  if (ci.comm_old_rank == 0)
+    compare_comm_old_and_cart_cartesian_mapping(
+      &cp, checkpoint_mapping, restart_mapping, comm_old_ranks_order,
+      comm_cart_ranks_order);
+
+  retval = MPI_Barrier(MPI_COMM_WORLD);
+  JASSERT(retval == MPI_SUCCESS).Text("MPI_Barrier(2) failed.");
+
+  retval = NEXT_FUNC(Bcast)(comm_old_ranks_order, cp.comm_old_size, MPI_INT, 0,
+                            comm_old_prime);
+  JASSERT(retval == MPI_SUCCESS)
+    .Text("Failed to broadcast <comm_old_ranks_order> integer array.");
+
+  retval = MPI_Barrier(MPI_COMM_WORLD);
+  JASSERT(retval == MPI_SUCCESS).Text("MPI_Barrier(3) failed.");
+
+  retval = NEXT_FUNC(Bcast)(comm_cart_ranks_order, cp.comm_cart_size, MPI_INT,
+                            0, *comm_cart_prime);
+  JASSERT(retval == MPI_SUCCESS)
+    .Text("Failed to broadcast <comm_cart_ranks_order> integer array.");
+
+  retval = MPI_Barrier(MPI_COMM_WORLD);
+  JASSERT(retval == MPI_SUCCESS).Text("MPI_Barrier(4) failed.");
+
+  // Create <comm_old> and <comm_cart> communicators
+  create_comm_old_communicator(&cp, comm_old_ranks_order);
+  create_comm_cart_communicator(&cp, comm_cart_ranks_order);
+
+  retval = MPI_Barrier(MPI_COMM_WORLD);
+  JASSERT(retval == MPI_SUCCESS).Text("MPI_Barrier(5) failed.");
+
+  // --------------------------------------------------------------------------
+
+  // Update mapping
   MPI_Comm virtComm = rec.args(5);
-  UPDATE_COMM_MAP(virtComm, *comm_cart_prime);
+  UPDATE_COMM_MAP(MPI_COMM_WORLD, comm_old);
```

Comment:
Add comment.
FIXME: This works for MPICH but maybe not for other MPI libraries.